### DISCUSSION TITLE
Capture __RequestVerificationCookie from response

### DISCUSSION THREAD

**Discussion by BanksySan** [artilleryio/artillery#1910] [Artillery]
I'm trying to capture this cookie, in one of the `Set-Cookie` headers on a response so that I can plus it into a form submission, however, the `capture` only seems to be able to take data from the payload of the response, not the headers.

How do I capture the `__RequestVerificationToken` cookie from the header?

**Comment by hassy** [comment]
hi @BanksySan 👋 if there are multiple `set-cookie` headers the best way is to do it with an `afterResponse` hook (otherwise you could use `capture` with a `header` option to grab the value of `set-cookie`).

https://www.artillery.io/docs/guides/guides/http-reference#afterresponse-hooks

The `res.headers` object will give you all of the headers.

You can set a scenario variable by adding a property to `context.vars` object, so e.g. setting this inside an `afterResponse` hook:

```
context.vars.myVar = 123;
```

Makes `myVar` usable in subsequent steps in the scenario, e.g.:

```
- get:
    url: "/foo/{{ myVar }}"
```

**Reply by BanksySan** [reply]
Thanks @hassy.  I'll go in that direction and stop fighting with the `capture` instructions.

**Reply by BanksySan** [reply]
The following code works:

```javascript
/**
 * Get the header value.
 * (Required as headers a case-insensitive)
 * @param headers {Object}
 * @param header {string}
 * @returns {undefined|*}
 */
function getHeader(headers, header) {
    const lowerPropertyName = header.toLowerCase();

    for (let key in headers) {
        if (key.toLowerCase() === lowerPropertyName) {
            return headers[key];
        }
    }

    return undefined;
}

/**
 * After response event handler.  Extract the <code>__RequestVerificationToken</code> from the <code>Set-Cookie</code> header.
 * @param requestParams {object}
 * @param response {IncomingMessage}
 * @param context {object}
 * @param ee {EventEmitter}
 * @param next {Function}
 */
module.exports.extractRequestVerificationToken = function extractRequestVerificationToken(requestParams, response, context, ee, next) {
    const EXTRACT_TOKEN_REGEX = /^__RequestVerificationToken=(?<token>[^;]*);[^\n\r]*$/
    const setCookieHeader = getHeader(response.headers, 'Set-Cookie');
    if (setCookieHeader) {
        for (let i = 0; i < setCookieHeader.length; i++) {
            const token = EXTRACT_TOKEN_REGEX.exec(setCookieHeader[i])?.groups['token'];
            if (token) {
                context.vars.csrf = token;
                break;
            }
        }
    }
    next();
}
```

Is there a way to supply the tag name that we want to use though?  I don't really like having it hard coded to `csrf`.
